### PR TITLE
Feature/swarm mode

### PR DESCRIPTION
**Pull request by jkp85**
Swarm mode implementation.
I managed to overcome ports issue (discussed in slack), so we can spawn as many services as we like on single swarm cluster. It was possible thanks to Traefik @roachmd suggested.
In order for this to work several steps are needed to be done before launch:
1. Create swarm mode cluster `docker swarm init`
2. Create global overlay network that Traefik needs (`docker network create --driver=overlay <net_name>`). Name of the network should be set in env file as `DOCKER_NET` var.
3. Traefik service should be created in swarm as described here: https://docs.traefik.io/user-guide/swarm-mode/#deploy-trfik
4. Remember to update openresty image that contains this PR: https://github.com/3Blades/openresty/pull/43

I wrote tests for critical parts of this implementation. Tests are creating real swarm cluster and spawn real services.

### CODE REVIEW

File: servers/spawners/base.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+from utils import create_jwt_token
+
+logger = logging.getLogger(__name__)
+
+
+class ServerSpawner(object):
+    """
+    Server service interface to allow start/stop/terminate servers
+    """
+    __metaclass__ = abc.ABCMeta
```
Comment (johngriebel) [inline servers/spawners/base.py]: FYI, `__metaclass__` is [ignored](http://python-3-patterns-idioms-test.readthedocs.io/en/latest/Metaprogramming.html#the-metaclass-hook-in-python-3) in Python 3.
You could change this to `class ServerSpawner(object, metaclass=abc.ABCMeta):`

Comment (jkp85) [inline servers/spawners/base.py]: Good catch, fixing :)

**Comment by johngriebel** [comment]
Looks good

**Comment by jgwerner** [comment]
@jkp85 could you clarify what states will be displayed? With Swarm Mode there is not `Stopped` concept, so we would need to eliminate those elements from CLI Tools and UI, correct? If so, I'll open new downstream issues.

**Comment by jkp85** [comment]
@jgwerner Yes, stop won't be present. There is only "Pending", "Running" and "Error"

File: servers/spawners/swarm_mode.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            name=self.server.container_name,
+            labels={'traefik.port': self._get_server_port()},
+            networks=[settings.DOCKER_NET, self.network_name],
+            resources=self._resources(),
+            restart_policy=self._restart_policy(),
+            update_config=self._update_config(),
+        )
+
+    def stop(self):
+        self.terminate()  # you can't stop a service
```
Comment (jgwerner) [inline servers/spawners/swarm_mode.py]: Wouldn't it be better to deprecate function?

Comment (jkp85) [inline servers/spawners/swarm_mode.py]: It could, I'll change that.

**Comment by johngriebel** [comment]
Linking this PR to issue #89.

**Comment by johngriebel** [comment]
@roachmd - We're waiting on your review for this. Can we go ahead and merge?

**Comment by roachmd** [comment]
I can't approve this until this is been tested with swarm mode

**Comment by jkp85** [comment]
Things needed for testing.

1. Working api setup (with docker-compose)
2. Swarm mode cluster (it shouldn't be the same that api is running on)
3. `3blades/app-backend` image build with code from this PR
    - clone my fork
    - checkout `feature/swarm_mode` branch
    - build an image with docker-compose `docker-compose build api celery`
4. `3blades/openresty` image from my fork (master branch)
    - clone my fork
    - build an image `docker build -t 3blades/openresty .`
5. Change `DOCKER_HOST`, `DOCKER_DOMAIN` and `DOCKER_PORT` env variables to point to swarm mode cluster.
6. Start api via docker-compose.
7. (Optional) Go to `http://<api_host>/tbs-admin/sites/site/` and change site host that to one that will be available from inside container (usually your ip, my is 192.168.0.100) and port. It should look like this 192.168.0.100:5000. This is necessary only if you want that server endpoints work out of the box.
8. Create a server.
9. Open server endpoint returned in server json.
Do some testing.

**Comment by roachmd** [comment]
Okay, To test I'm using performing those steps a local Vagrant setup

1) I will use ondemo to build a working api locally
2) Build a local swarm cluster with Traefik.
3) Rebuild (Openresty, app-backend) 
4) Down and up with docker compose.

**Comment by roachmd** [comment]
@jkp85 I'm testing the steps necessary to update On premise, I will follow up by testing a two node cluster.

**Comment by jgwerner** [comment]
@roachmd how is this looking?

**Comment by roachmd** [comment]
@jgwerner docker swarm init produces a tls. Struggling with the docks on how not to make that happen

**Comment by jgwerner** [comment]
@roachmd @jkp85 @johngriebel if we get stick disabling TLS what would be the cost of refactor to leave TLS enabled?

**Comment by johngriebel** [comment]
@jgwerner I'm not sure, that would be a question specifically for @jkp85. However, my guess is that it would be non-trivial.

**Comment by jkp85** [comment]
@jgwerner @roachmd Enabling tls should be resolved by config. In all repositories I'm creating docker client from env variables. Config requires for all clients to have access to docker daemon certificates with files like specified here: https://github.com/docker/docker-py/blob/master/docker/utils/utils.py#L444 and here https://github.com/moby/moby/blob/master/client/client.go#L116 Three envs that are needed are: `DOCKER_HOST`, `DOCKER_TLS_VERIFY=1`, `DOCKER_CERT_PATH`.

**Comment by roachmd** [comment]
@jgwerner We had a some success this morning. At this time, I believe we should'nt need to refactor.

**Comment by jgwerner** [comment]
@johngriebel @jkp85 @roachmd do we need to keep this PR open?

**Comment by jkp85** [comment]
@roachmd I agree. Closing.

**Pull request closed without merging**